### ISSUE TITLE
NeRF instance segmentation

### ISSUE THREAD

**Issue by Lizhinwafu** [THU-luvision/OmniSeg3D#4]
When i use this command CUDA_VISIBLE_DEVICES=0 opt=show_sem  bash scripts/run_llff.sh, I have this error.
![image](https://github.com/user-attachments/assets/50c3d7bf-f356-4fdc-b4c0-79fec13960f5)

**Comment by OceanYing** [comment]
This error indicates the addmm_impl_cpu module in torch.matmul() function does not support input tensor with 16bit (half float) on CPU. You can try to move tensor 'X' onto GPU to suppress the error. Or you may try to modify the data type of tensor 'X' instead:
```
# On CPU
covariance_matrix = (1 / n) * torch.matmul((X.to(torch.float32)).T, X.to(torch.float32)).float()
```

**Comment by Lizhinwafu** [comment]
![image](https://github.com/user-attachments/assets/c93e4f48-55fd-425f-81ac-24fdaea0f319)
My target plant is a peanut, but after training, there is fog around the plant, and the rendered image is also blocked by the fog. What is the reason and how to solve it?

**Comment by Lizhinwafu** [comment]
![image](https://github.com/user-attachments/assets/30a30096-e4d6-4f69-9acb-0a188b620829)
My goal is to segment the fruit instances. Can omniseg3d do this? Where is the segmentation result saved?

**Comment by OceanYing** [comment]
This InstantNGP-based version support segment object on multi-view images (2D). To segment object out explicitly in 3D space, I would suggest you use our [OmniSeg3D-GS](https://github.com/OceanYing/OmniSeg3D-GS), which is implemented based on 3D gaussian splatting. Since 3DGS is a point-based representation, chosen objects can be easily segmented out in our GUI. Please check the link of [GUI](https://github.com/OceanYing/OmniSeg3D-GS?tab=readme-ov-file#gui-visualization-and-segmentation) to find the instruction of how to segment objects and save them as .ply file.

**Comment by OceanYing** [comment]
> ![image](https://private-user-images.githubusercontent.com/33644565/375588912-c93e4f48-55fd-425f-81ac-24fdaea0f319.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MTAyOTYsIm5iZiI6MTcyODYwOTk5NiwicGF0aCI6Ii8zMzY0NDU2NS8zNzU1ODg5MTItYzkzZTRmNDgtNTVmZC00MjVmLTgxYWMtMjRmZGFlYTBmMzE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDAxMjYzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjZjEwYWQ1MTA4ZWM0YjkyMTk3MDBmZWEwMDFmZTc0NjlkY2E5NGVmZGE2ZWZkYTZlY2Y5Y2ExMTU3Yzk2MWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tTdEbc6yphD_LAC8Hf9thKslXMkBSg05IUylXe-tDMs) My target plant is a peanut, but after training, there is fog around the plant, and the rendered image is also blocked by the fog. What is the reason and how to solve it?

Here are several possible reasons:
1. The scene capture: If the observation is too sparse, the performance (especially the reconstructed geometry) of Instant-NGP may drop down and show such floaters. So you can try to obtain more images to alleviate this phenomenon.
2. The boundary setting: Instant-NGP requires a good bounding box that cover the target object. You can check the bounding box to see if there are any deviation to be fixed.
3. The reconstruction pipeline may not be powerful enough. You can try to use 3DGS as the reconstruction baseline, which may give more robust performance for rendering. (check [OmniSeg3D-GS](https://github.com/OceanYing/OmniSeg3D-GS)).

**Comment by Lizhinwafu** [comment]
![image](https://github.com/user-attachments/assets/c0d2956a-cc7a-4b2e-be05-787e548d3753)
I give a point prompt on my fruit, maybe it can't segment the fruit. 

For 3DGS, It can segment the fruit with a good edge? I remember that the point cloud of 3DGS cannot segment the edges very well.

**Comment by Lizhinwafu** [comment]
I want to perform instance segmentation and count my fruits, do you have any recommended methods in NeRF/3DGS? Because my fruits are touching each other, I think 3D point cloud segmentation network will not give good results.

**Comment by Lizhinwafu** [comment]
I used the training set in llff format and used SAM to get the 2D mask of the image. I am wondering if I can use the 2D instance segmentation model to directly get the 2D mask of the fruit on the image to train omniseg3d (only using the 2D instance mask of the fruit)? Can I get the 3D instance fruit after training? Because I want to remove the step of manually giving prompts.

**Comment by OceanYing** [comment]
> ![image](https://private-user-images.githubusercontent.com/33644565/375594444-c0d2956a-cc7a-4b2e-be05-787e548d3753.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MTEwNDAsIm5iZiI6MTcyODYxMDc0MCwicGF0aCI6Ii8zMzY0NDU2NS8zNzU1OTQ0NDQtYzBkMjk1NmEtY2M3YS00YjJlLWJlMDUtNzg3ZTU0OGQzNzUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDAxMzkwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEyOWFmYTYyYWViZDgzNDEzNTEzZGIwNjNkZTI2Mzk4MDI2M2NmNzI3NWUwNTg3M2VkZGY4M2E2ZDYwOTNhMTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pBKRdrGQQX9NpWjqAV25hYosNDS-lRyVkKN5ywZyKfE) I give a point prompt on my fruit, maybe it can't segment the fruit.
> 
> For 3DGS, It can segment the fruit with a good edge? I remember that the point cloud of 3DGS cannot segment the edges very well.

### 1. For your segmentation results:
Hi, would you please share more details about how you give the point prompt and get the binary segmentation image? And just curious about it: Where is the red rectangle come from? Since we just click on the target object in the GUI screen (where only single pixel is selected) but no binary output is provided directly.

In the preprocessing stage, you can check the output masks and patches from SAM. If the target object cannot be segmented out in SAM, it cannot be done by OmniSeg3D as well.

If SAM can segmented out the fruits, you can check the rendered semantic feature image to see if the object has different feature colors. If so, I think the individual object can be segmented. From my experiments, OmniSeg3D can segment object at different graininess successfully in most cases.

### 2. For 3DGS:
I think 3DGS can also provide good edge from my experiments.

**Comment by OceanYing** [comment]
> I used the training set in llff format and used SAM to get the 2D mask of the image. I am wondering if I can use the 2D instance segmentation model to directly get the 2D mask of the fruit on the image to train omniseg3d (only using the 2D instance mask of the fruit)? Can I get the 3D instance fruit after training? Because I want to remove the step of manually giving prompts.

I think it is doable but still cannot prevent the manual labeling operation, since you need to label/choose the mask of target objects for each segmented image, which may cost even more manpower.

**Comment by Lizhinwafu** [comment]
![025_s](https://github.com/user-attachments/assets/d31155aa-de01-4a62-9e70-9d5dbcfbb47f)
This is the automatically saved segmentation result. The red box is added manually, indicating that I give a prompt on this fruit.

**Comment by OceanYing** [comment]
> I want to perform instance segmentation and count my fruits, do you have any recommended methods in NeRF/3DGS? Because my fruits are touching each other, I think 3D point cloud segmentation network will not give good results.

I agree that general learning-based 3D point cloud segmentation methods may not give promising results since the network may not have seen the specific object class before. 

On the other hand, since you want to get the instance segmentation of fruit only, this goal naturally contains some semantic level requirement for the segmentation method you want to implement. To this end, some extra information should be provided, such as: the target object may have the similar physical size or shape. I think this is a topic worth exploring, which is how to specify the target object for a class-agnostic 3D segmentation algorithm.

**Comment by Lizhinwafu** [comment]
Thanks, I will test omniseg3d-gs first.

**Comment by OceanYing** [comment]
> ![025_s](https://private-user-images.githubusercontent.com/33644565/375600748-d31155aa-de01-4a62-9e70-9d5dbcfbb47f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MTI1MDQsIm5iZiI6MTcyODYxMjIwNCwicGF0aCI6Ii8zMzY0NDU2NS8zNzU2MDA3NDgtZDMxMTU1YWEtZGUwMS00YTYyLTllNzAtOWQ1ZGJjZmJiNDdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDAyMDMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVkZDY5NGQyMDVhNDBjNTg2MjUzYzk0YWU2ZWY5MTEzZWVkNGM0ODVmNzYyNzU1Y2IzNmQ2ZTJhMmE4YmViZDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EfFq-4cIpn4P20iztvsS2ZyF65FWZZ1ONIyVgzqICyw) This is the automatically saved segmentation result. The red box is added manually, indicating that I give a prompt on this fruit.

This looks cool! The semantic feature map looks normal. May I know how you **give a prompt**? And **what** you give the prompt to? How do you get the output binary mask? 

If you are trying to do the segmentation on a rendered 2D feature map, I suggest to select just one pixel of the target object as prompt, not a square region. Another way is using our GUI to do the segmentation. NOTE: Both schemes require tuning the threshold

**Comment by Lizhinwafu** [comment]
![image](https://github.com/user-attachments/assets/5e612591-264c-4bac-83d9-fac401be27c7)
I right-clicked the mouse to give a hint point, and adjusted ScoreThr and XYZThr to the maximum, giving a fruit mask.

**Comment by Lizhinwafu** [comment]
NOTE: Both schemes require tuning the threshold
I used the default thresholds in the code and did not adjust them. Which thresholds need to be adjusted?

One of the problems is that my target is surrounded by a fog, and I need to slowly find my target. I also encountered this problem when training SA3D. But I don't have this problem when training on nerfstudio.

I previously had an operation that removed the manual prompting method. When I was using the SAGA model, I first used YOLOv8 to obtain a 2D mask of a fruit, and used the pixels corresponding to the entire mask as prompts for the next step. If I want to use the same operation, which code needs to be changed?

**Comment by OceanYing** [comment]
> ![image](https://private-user-images.githubusercontent.com/33644565/375604084-5e612591-264c-4bac-83d9-fac401be27c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg2MTUwNzgsIm5iZiI6MTcyODYxNDc3OCwicGF0aCI6Ii8zMzY0NDU2NS8zNzU2MDQwODQtNWU2MTI1OTEtMjY0Yy00YmFjLTgzZDktZmFjNDAxYmUyN2M3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDExVDAyNDYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2OWNmMWE4Mzc0Yzc4ODJlMjJkYzM5NDE3MTI0NTAzNWFhM2I3YWZhZDY4ZWRjNDc5Mzc4NjAwZTY1MzRjMTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5xmXOHWpstcv386PKrIoctlanlKPpt8arr_MbNWtn4E) I right-clicked the mouse to give a hint point, and adjusted ScoreThr and XYZThr to the maximum, giving a fruit mask.

I think you don't have to **adjust both ScoreThr and XYZThr to the maximum**. It would be helpful to have a look at our paper to see the definition of these thresholds.

In fact, you need to find good values for both ScoreThr and XYZThr, which indicates a good segmentation, then you turn on the "binary threshold" on the right panel to see the binary segmentation results.

**Comment by OceanYing** [comment]
> NOTE: Both schemes require tuning the threshold I used the default thresholds in the code and did not adjust them. Which thresholds need to be adjusted?
> 
> One of the problems is that my target is surrounded by a fog, and I need to slowly find my target. I also encountered this problem when training SA3D. But I don't have this problem when training on nerfstudio.
> 
> I previously had an operation that removed the manual prompting method. When I was using the SAGA model, I first used YOLOv8 to obtain a 2D mask of a fruit, and used the pixels corresponding to the entire mask as prompts for the next step. If I want to use the same operation, which code needs to be changed?

1. I mean the thresholds in the right panel of the GUI. See [this](https://github.com/THU-luvision/OmniSeg3D/issues/4#issuecomment-2406452180) for more information.
2. I think that's because the Instant-NGP backbone is not robust enough. I use [NGP_pl](https://github.com/kwea123/ngp_pl) as our backbone, which may show worse reconstruction performance than the one in nerfstudio.
3. You should change the output of `run_sam.py`, where the `indices` of unchosen object should be set to 0 while the indices value of chosen object just remain unchanged.

**Comment by Lizhinwafu** [comment]
Thanks, I will test.
